google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/StreamingSubscriberConnection.java

```diff
@@ -0,0 +1,219 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import com.google.auth.Credentials;
+import com.google.cloud.Clock;
+import com.google.cloud.pubsub.Subscriber.MessageReceiver;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.pubsub.v1.StreamingPullRequest;
+import com.google.pubsub.v1.StreamingPullResponse;
+import com.google.pubsub.v1.SubscriberGrpc;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.Status;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientCalls;
+import io.grpc.stub.ClientResponseObserver;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implementation of {@link AbstractSubscriberConnection} based on Cloud Pub/Sub streaming pull. */
+final class StreamingSubscriberConnection extends AbstractSubscriberConnection {
+  private static final Logger logger = LoggerFactory.getLogger(StreamingSubscriberConnection.class);
+
+  private static final Duration INITIAL_CHANNEL_RECONNECT_BACKOFF = new Duration(100); // 100ms
+  private static final int MAX_PER_REQUEST_CHANGES = 10000;
+
+  private Duration channelReconnectBackoff = INITIAL_CHANNEL_RECONNECT_BACKOFF;
+
+  private final Channel channel;
+  private final Credentials credentials;
+
+  private ClientCallStreamObserver<StreamingPullRequest> requestObserver;
+
+  public StreamingSubscriberConnection(
+      String subscription,
+      Credentials credentials,
+      MessageReceiver receiver,
+      Duration ackExpirationPadding,
+      int streamAckDeadlineSeconds,
+      Distribution ackLatencyDistribution,
+      Channel channel,
+      FlowController flowController,
+      ScheduledExecutorService executor,
+      Clock clock) {
+    super(
+        subscription,
+        receiver,
+        ackExpirationPadding,
+        ackLatencyDistribution,
+        flowController,
+        executor,
+        clock);
+    this.credentials = credentials;
+    this.channel = channel;
+    setMessageDeadlineSeconds(streamAckDeadlineSeconds);
+  }
+
+  @Override
+  protected void doStop() {
+    super.doStop();
+    requestObserver.onError(Status.CANCELLED.asException());
+  }
+
+  @Override
+  void initialize() {
+    final SettableFuture<Void> errorFuture = SettableFuture.create();
+    final ClientResponseObserver<StreamingPullRequest, StreamingPullResponse> responseObserver =
```

Comment:
Done. 

Honestly to me this one didn't look like a heavy-duty anonymous class (I know there is no hard rule on this and sometimes is just a matter of the dev taste) but I'm happy to comply :).

Let me know if you want me to also pull out the anonymous FutureCallback class below as well. But again I feel that this is one is also simple enough to read inline.